### PR TITLE
overlord: export and use snapd tools

### PR DESCRIPTION
**Pull request by zyga**
This branch addresses a long-standing issue, where snapd tools provided
by snapd snap were not kept up-to-date, as seen from the perspective of
running snap applications. This is because snapd tools were bind-mounted
from whatever was providing snap-confine at the time the snap
application process, for a given snap, was first started and there was
no mechanism that would keep it up-to-date as time goes on.

We have discussed this problem for a while and picked solution "C" as
described on the referenced Snapcraft Forum. The solution does not model
the dependency between all snaps and snapd or core at the level of snap
interfaces. Instead snapd provides a directory, accessible from all
participating mount namespaces, where up-to-date tools are available,
and uses those instead. The tools are provided as symbolic links through
a more generic export mechanism, which lives in a new overlord manager.

The export manager allows any installed revision to export content in
the form of symbolic links. Manifest of the content exported by a
specific revision is stored in snap state. The manager introduces two
new tasks, export-content and unexport-content, which are responsible
for the bulk of the work of putting the symbolic links in place or
removing them when no longer necessary.

The tasks are now emitted by install, remove, enable and disable
operations. They are ordered such, so that export-content is always
before link-snap, and unexport-content is always after link snap.

Out of all the revisions of a given snap, or in more general terms, out
of the available exported versions, a single version is selected as
current. Most of the time the exported version is the active revision of
a given snap.

Since snapd tools are a special, and can be provided by the classic host
system package, any revision of core and any revision of snapd
simultaneously, some more complexity arises when selecting the current
provider of snapd tools. On core systems snapd is preferred over core.
On classic systems host tools are used in absence of other snaps or
when re-execution is disabled.

In all cases a new kind of "current" symlink is created, which points to
the directory containing content exported by the picked provider.

The export manager updates the current symlink each time link-snap or
unlink-snap tasks, from the snap manager, are done or undone. This
arrangement was picked due to important synchronization role of
link-snap and due to the difficulty of introducing additional task which
runs immediately after link and unlink tasks in all cases.

On startup, the export manager ensures that snapd tools from all sources
are exported and that the current symlink is up-to-date and in place.
This ensures that before any other manager operates, such as when the
interface manager chooses to re-generate profiles, snapd tools are
available from the fixed location: /var/lib/snapd/export/snapd/current/tools.

Snap confine is updated to use the tools from the new location, but as a
safety net, a new experimental feature is introduced. This feature,
unlike before, is enabled by default. Ubuntu Core customers experiencing
unforeseen issues could use their management agents to disable this
feature, making snap-confine bind-mount tools as we did before. This
feature does not affect the export process, which is relatively
harmless, only the usage of the exported tools.

A new spread test is examining the fine details of the export behavior.
Both the new default and the old methods are tested. The test invariant
is also extended to, at all times, ensure that snapd tools are exported
correctly. Some unusual system and test behavior is also documented there.

The apparmor profiles used by snapd were extended to cover both the new
and the old behavior. Some minor inconsistency about the use of lib64
was also corrected, to ensure that less often used file system hierarchy
variants are supported.

The mount profile test has been updated to reflect the mount changes
required to use the exported tools. At runtime, snap-confine mounts
a tmpfs at /usr/lib/snapd, as seen inside them per-snap mount namespace,
and puts symlinks to all the tools that are used from such namespace
there.

Ref: https://forum.snapcraft.io/t/12139

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by zyga** [comment]
I pushed SELinux changes now.

File: overlord/exportstate/manifest.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+// used, if it returns false PathInSnapMountNS is used instead.
+//
+// This distinction enables exporting files that are consumed by either other
+// snaps or by the classic system.
+type exportEntry struct {
+	PathInExportSet   string
+	PathInHostMountNS string
+	PathInSnapMountNS string
+
+	IsExportedPathValidInHostMountNS bool
```
Comment (pedronis) [inline overlord/exportstate/manifest.go:191]: I see this read but I don't understand where it is set, is not set below for example

**Review by pedronis — COMMENTED** [review]
Changes look good, we can discuss details further in the smaller PRs. One question though.

File: overlord/exportstate/manifest.go (hunk 1 above)
Comment (zyga) [reply overlord/exportstate/manifest.go:191]: It's not set to true in any of the use cases we have _yet_. It's there to remind the reader that there is a difference and not all symlinks are valid at all times.

**Comment by zyga** [comment]
I've split link-participant out of this pull request as https://github.com/snapcore/snapd/pull/9422 

Once it is merged I will rebase this pull request on top of that change.

**Comment by zyga** [comment]
Rebased on merged link-participant changes.

**Comment by zyga** [comment]
I've split off https://github.com/snapcore/snapd/pull/9546 as the first fragment. I will rebase this branch on top shortly to facilitate iteration.

**Comment by zyga** [comment]
I've rebased this again to reflect all the changes to the export manager. No changes to the tip patch were required.

**Pull request closed without merging**